### PR TITLE
[Inference/Kernel]Add get_cos_and_sin Kernel

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
pytest:
image

**Model Benchmark:**
***Total Time Comparison***

| bsz | in_len | out_len | Throughput (tokens/sec) | 
|-----|--------|---------|--------------------------|
| 16   | 128    | 128     | 1793.14 -> 1806.67 | 
| 32  | 128    | 128     | 3099.44 -> 3112.40 | 
| 64  | 128    | 128     | 5006.11 -> 5010.27 | 
| 16   | 512    | 128     | 1394.90 -> 1407.57 | 
| 32  | 512    | 128     | 2060.37 -> 2081.11 | 
| 64  | 512    | 128     | 2746.49 -> 2748.17 | 

**OP Benchmark（batchsize 16，Input len 128，output len 15）:**
***OP CPU Time Comparison:***

| cuda | triton |
| :-----| ----: |
| 0.039ms | 0.274ms |

triton:
image
cuda:
image
***OP GPU Time Comparison:***

| cuda | triton |
| :-----| ----: |
| 0.067ms | 0.283ms |

triton:
image
cuda:
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        if (is_prompts){                                                                                            \
+            apply_get_context_cos_and_sin_kernel<scalar_t, __aligned, __vec_size><<<grid, block, 0, stream>>>(      \
+                cos.data_ptr<scalar_t>(),                                                                           \
+                sin.data_ptr<scalar_t>(),                                                                           \
+                cos_cache.data_ptr<scalar_t>(),                                                                     \
+                sin_cache.data_ptr<scalar_t>(),                                                                     \
+                sequence_lengths.data_ptr<int>(),                                                                \
+                comsum_lengths.data_ptr<int>(),                                                                \
+                batch_size,                                                                                         \
+                head_dim                                                                                         \
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/get_cos_and_sin_kernel.cu:127]: Trivial: formatting

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    at::Tensor& sin,                 // [num_tokens, head_dim]
+    at::Tensor& sequence_lengths,    // [batch_size]
+    int max_seq_len_in_batch,
+    bool is_prompts
+) {
+    int token_num = cos.size(0);
+    int head_dim = cos.size(1);
+    int batch_size = sequence_lengths.size(0);
+
+    at::Tensor comsum_lengths;
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/get_cos_and_sin_kernel.cu:87]: Typo: replace all `comsum`

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        copy_vector<scalar_t, VecSize>(sin + dest_offset_id + begin_id, sin_cache_ptr + src_offset_id + begin_id);
+    }
+
+    if (!Aligned) {
+        for (; begin_id < head_dim; ++begin_id ) {
+            cos[dest_offset_id + begin_id] = cos_cache_ptr[src_offset_id + begin_id];
+            sin[dest_offset_id + begin_id] = sin_cache_ptr[src_offset_id + begin_id];
+        }
+    }
+}
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/get_cos_and_sin_kernel.cu:71]: Can this two be merged to a general get_cos_sin kernel function to reduce redundancy?

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,192 @@
+#include <ATen/cuda/CUDAContext.h>
+#include <torch/extension.h>
+
+#include "utils/vector_copy_utils.h"
+#include "../common/micros.h"
+#include "stdio.h"
+
+template <typename scalar_t, bool Aligned, int VecSize>
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/get_cos_and_sin_kernel.cu:8]: Just wonder why not putting `Aligned` and `VecSize` as function parameters ,instead of template parameters? Using a non-manifestly-constant-evaluated expression as a non-type template parameter is not recommended. This is actually causing potential issues we discussed before in this [PR](https://github.com/hpcaitech/ColossalAI/pull/5461).

**Review by yuanheng-zhao — CHANGES_REQUESTED** [review]

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu (hunk 4 above)
Comment (yuehuayingxueluo) [reply extensions/csrc/cuda/get_cos_and_sin_kernel.cu:8]: When performing vectorized reading, it is necessary to pass a constant, thus the VecSize needs to be passed through templates.

Comment (yuehuayingxueluo) [reply extensions/csrc/cuda/get_cos_and_sin_kernel.cu:8]: Passing a bool value through a template allows the CUDA compiler to optimize ‘if operations during compilation.

**Review by yuanheng-zhao — APPROVED** [review]

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu (hunk 3 above)
Comment (yuehuayingxueluo) [reply extensions/csrc/cuda/get_cos_and_sin_kernel.cu]: Done

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu (hunk 2 above)
Comment (yuehuayingxueluo) [reply extensions/csrc/cuda/get_cos_and_sin_kernel.cu:110]: Done

File: extensions/csrc/cuda/get_cos_and_sin_kernel.cu (hunk 1 above)
Comment (yuehuayingxueluo) [reply extensions/csrc/cuda/get_cos_and_sin_kernel.cu]: Done

File: tests/test_infer/test_ops/cuda/test_get_cos_and_sin.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+from tests.test_infer.test_ops.triton.test_xine_copy import get_cos_sin
+
+inference_ops = InferenceOpsLoader().load()
+
+
+def numpy_equal(x, y):
+    x_numpy = x.detach().cpu().numpy()
+    y_numpy = y.detach().cpu().numpy()
+
+    np.testing.assert_allclose(x_numpy, y_numpy)
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_ops/cuda/test_get_cos_and_sin.py:15]: This seems to be `numpy_allclose`

Comment (yuanheng-zhao) [reply tests/test_infer/test_ops/cuda/test_get_cos_and_sin.py:15]: Is it possible to use metric of "equivalence" to compare kernels of getting cos/sin cache?

Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/cuda/test_get_cos_and_sin.py]: Done
